In [1]:
# import packages
import keras
from keras import layers
from keras import models
from keras.models import Sequential
from keras.layers import Conv2D, Activation, MaxPooling2D, Flatten, Dense, AveragePooling2D

# checking the version of keras
print(keras.__version__)

Using TensorFlow backend.


2.2.4


In [2]:
# fully connected layer only
model = Sequential([
    Dense(100,input_shape = (28,28,1)),
    Activation('relu'),
    Dense(10),
    Activation('softmax'),
])
model.summary()

W0806 12:41:08.479124 140051457070976 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:74: The name tf.get_default_graph is deprecated. Please use tf.compat.v1.get_default_graph instead.

W0806 12:41:08.532277 140051457070976 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:517: The name tf.placeholder is deprecated. Please use tf.compat.v1.placeholder instead.

W0806 12:41:08.540390 140051457070976 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:4138: The name tf.random_uniform is deprecated. Please use tf.random.uniform instead.



_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 28, 28, 100)       200       
_________________________________________________________________
activation_1 (Activation)    (None, 28, 28, 100)       0         
_________________________________________________________________
dense_2 (Dense)              (None, 28, 28, 10)        1010      
_________________________________________________________________
activation_2 (Activation)    (None, 28, 28, 10)        0         
Total params: 1,210
Trainable params: 1,210
Non-trainable params: 0
_________________________________________________________________


In [3]:
# two convolution layers with fully connected layer (NO Pooling layer)
model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), input_shape=(28, 28, 1)))
model.add(layers.Conv2D(25, (3, 3)))

model.add(Flatten())
model.add(Dense(units=100))
model.add(Activation('relu'))

model.add(Dense(units=10))
model.add(Activation('softmax'))

model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 24, 24, 25)        7225      
_________________________________________________________________
flatten_1 (Flatten)          (None, 14400)             0         
_________________________________________________________________
dense_3 (Dense)              (None, 100)               1440100   
_________________________________________________________________
activation_3 (Activation)    (None, 100)               0         
_________________________________________________________________
dense_4 (Dense)              (None, 10)                1010      
_________________________________________________________________
activation_4 (Activation)    (None, 10)                0         
Total para

In [4]:
# two convolution layers with fully connected layer (with Pooling layer)
model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), input_shape=(28, 28, 1)))
model.add(AveragePooling2D(2,2))
model.add(layers.Conv2D(25, (3, 3)))
model.add(AveragePooling2D(2,2))

model.add(Flatten())
model.add(Dense(units=100))
model.add(Activation('relu'))

model.add(Dense(units=10))
model.add(Activation('softmax'))

model.summary()

W0806 12:41:08.763418 140051457070976 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:3980: The name tf.nn.avg_pool is deprecated. Please use tf.nn.avg_pool2d instead.



_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_3 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
average_pooling2d_1 (Average (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_4 (Conv2D)            (None, 11, 11, 25)        7225      
_________________________________________________________________
average_pooling2d_2 (Average (None, 5, 5, 25)          0         
_________________________________________________________________
flatten_2 (Flatten)          (None, 625)               0         
_________________________________________________________________
dense_5 (Dense)              (None, 100)               62600     
_________________________________________________________________
activation_5 (Activation)    (None, 100)               0         
__________

When building a fully-connected layer with convolution layers before, the total parameters would be: 1,210

But when adding two convolution layers in the begining, the total parameters increased!!! (1,448,655 in total)

And, when adding averagePooling2D after each convolution layer, the total parameters decreased( 71,155 in total)